File: swift/leet_code/402.swift
/**
 `Solution` class contains the solution for the problem.

 This class has a single method `removeKdigits` which is used to remove `k` digits from the input `num` to get the smallest possible number.
 */
class Solution {
    /**
      This function removes `k` digits from the input `num` to get the smallest possible number.

      - Parameters:
        - num: The original number represented as a string.
        - k: The number of digits to remove.
      - Returns: The smallest possible number after removing `k` digits. If all digits are removed, returns "0".

      The function works by converting the input string to an array of `CChar` for processing. It then iterates over each character in the number, and while there are still digits to remove and the last digit in the stack is greater than the current digit, it removes the last digit from the stack. If there are still digits to remove after this process, it removes them from the end. Finally, it finds the index of the first non-zero digit and returns the number as a string, or "0" if all digits are zero.
     */
    func removeKdigits(_ num: String, _ k: Int) -> String {
        // Convert the input string to an array of CChar for processing
        let na = num.utf8CString
        // Define a zero value for comparison
        let zv = "0".utf8CString
        // If the number of digits to remove is greater than or equal to the number of digits in the input, return "0"
        if k >= na.count - 1 { return "0" }
        // Initialize the number of digits to remove and the stack to hold the digits of the number
        var rem = k, st: [CChar] = []
        // Iterate over each character in the number
        for i in 0 ..< na.count - 1 {
            let e = na[i]
            // While there are still digits to remove and the last digit in the stack is greater than the current digit
            while !st.isEmpty, rem > 0, st.last! > e {
                // Remove the last digit from the stack
                st.removeLast()
                // Decrease the number of digits to remove
                rem -= 1
            }
            // Add the current digit to the stack
            st += [e]
        }
        // If there are still digits to remove, remove them from the end
        for i in 0 ..< rem {
            st.removeLast()
        }
        // Find the index of the first non-zero digit
        let j = st.firstIndex(where: { $0 != zv[0] }) ?? st.count
        // If the index is less than the count of digits in the stack, return the number as a string, otherwise return "0"
        return j < st.count ?st[j...].reduce(into: "") { $0 += String(UnicodeScalar(UInt8($1))) } : "0"
    }
}

File: typescript/leet_code/402.js
/**
 * @fileoverview This module contains a solution for the "Remove K Digits" problem from LeetCode.
 * The problem is solved by using a stack to keep track of the smallest number that can be formed.
 */
/**
 * Function to remove k digits from the number to make it smallest.
 * @param {string} num - The number from which digits are to be removed.
 * @param {number} k - The count of digits to be removed.
 * @return {string} - The smallest possible number after removing k digits.
 */
function removeKdigits(num, k) {
    // Initialize stack and removed digits count
    var stk = [], rem = 0;
    // Iterate over each digit in the number
    for (var _i = 0, num_1 = num; _i < num_1.length; _i++) {
        var n = num_1[_i];
        // While stack is not empty and top of stack is greater than current digit and removed digits count is less
        // than k
        while (stk.length && n < stk[stk.length - 1] && rem < k) {
            // Remove the top of stack
            stk.pop();
            // Increment the removed digits count
            rem++;
        }
        // Push the current digit to the stack
        stk.push(n);
    }
    // If there are still digits to remove
    while (rem < k) {
        // Remove the top of stack
        stk.pop();
        // Increment the removed digits count
        rem++;
    }
    // Remove leading zeros
    while (stk[0] === "0") {
        stk.shift();
    }
    // If stack is not empty, join the digits to form the number, else return "0"
    return stk.length ? stk.join("") : "0";
}
